benches/basename_bench.rs

```diff
@@ -0,0 +1,133 @@
+#![feature(test)]
+extern crate test;
+use test::Bencher;
+
+include!("../src/lib.rs");
+
+use basename::basename;
+use path_parsing::{find_last_sep_pos, find_last_non_sep_pos, find_last_dot_pos, find_last_word};
+use std::ops::Range;
+use extname::extname;
+use memrnchr::memrnchr;
+extern crate memchr;
+
+#[bench]
+fn basename_version1_benchmark(b: &mut Bencher) {
+  let parameters = [
+    ("/a/b///c//////", ""),
+    ("file.test.", ".*"),
+    ("", ""),
+    ("///", ""),
+    ("..", ""),
+    ("..", "."),
+    ("..", "...")
+  ];
+
+  b.iter(|| {
+    parameters.iter().for_each(|&(a,b)| {
+      basename_version1(a,b);
+    })
+  })
+}
+
+// BUG # 162
+pub fn basename_version1<'a>(path: &'a str, ext: &str) -> &'a str {
+  let bytes: &[u8] = path.as_bytes();
+  let mut left: usize = 0;
+  let mut right: usize = bytes.len();
+  if let Some(last_slash_pos) = find_last_sep_pos(bytes) {
+    if last_slash_pos == right - 1 {
+      if let Some(pos) = find_last_non_sep_pos(&bytes[..last_slash_pos]) {
+        right = pos + 1;
+      } else {
+        return "/";
+      }
+      if let Some(pos) = find_last_sep_pos(&bytes[..right]) {
+        left = pos + 1;
+      }
+    } else {
+      left = last_slash_pos + 1;
+    }
+  }
+  let ext_bytes = ext.as_bytes();
+  if ext_bytes == b".*" {
+    if let Some(dot_pos) = find_last_dot_pos(&bytes[left..right]) {
+      right = left + dot_pos;
```

Comment:
To fix the wrap, surround the line with `if right != left + dot_pos + 1`?

Reply:
I'll give it a try.  _My version doesn't work on via monkeypatch branch so even that needs to be changed._

Reply:
Which line?  Can you give me a number.

Reply:
nvm, I think I got it: # 55

Reply:
Nope, now that fails the trailing dot test

```
failures:

---- basename::trailing_dot stdout ----
	thread 'basename::trailing_dot' panicked at 'assertion failed: `(left == right)`
  left: `"file.test."`,
 right: `"file.test"`', src/basename.rs:49:3
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Reply:
I'll finish up my fix with my version for now.

Reply:
Nvm, that's not right, because:

```ruby
File.basename('.c.', '.*') => '.c'
```

Looks like there is only one special case: `..`, so that line should be very easy to fix.
Instead of my suggestion above, wrap the if at L54 with another condition:
```rust
if bytes[left..right] != b".."

```

Reply:
Aand, it's not that simple. I'll look into this later.

Reply:
FYI… I just got my version to pass the monkeypatch tests with the Ruby/Spec suite.